### PR TITLE
Add support for >8 character StuffIt 4 passwords.

### PR DESCRIPTION
**Pull request by siddhartha77**
Add support for >8 character StuffIt 4 passwords.

### CODE REVIEW

**Comment by PaulTaykalo** [comment]
Hi, thanks for the contribution!
Can you please, provide some examples of the archives with under 8 charactes and > 8 characters just to be used as the test cases? thanks

**Comment by siddhartha77** [comment]
[archives.zip](https://github.com/MacPaw/XADMaster/files/12894915/archives.zip)

Sure, here are two archives. You'll have to unzip them first as GitHub doesn't allow MacBinary uploads. The Unarchiver handles the MacBinary format like a charm. The passwords are the names of the files: 1234567 and 1234567890123456 respectively.

File: XADStuffItDESHandle.m [hunk 1]
```diff
@@ -38,19 +38,33 @@ +(NSData *)keyForPasswordData:(NSData *)passworddata entryKey:(NSData *)entrykey
...
+	uint32_t x = 0;
+	uint32_t length=[passworddata length];
+  
+	StuffItDESSetKey(archivekey,&ks);
+	
+	do
+ 	{
+		 uint8_t passblock[8] = { 0, 0, 0, 0, 0, 0, 0, 0 };
+
+		memcpy(passblock, [passworddata bytes] + (x << 3), 8);
```
Comment (PaulTaykalo) [inline XADStuffItDESHandle.m:54]: 🤔 
this could lead to the reading out of bounds, in case if password is not multiple of 8

**Review by PaulTaykalo — CHANGES_REQUESTED** [review]

File: XADStuffItDESHandle.m (hunk 1 above)
Comment (siddhartha77) [reply XADStuffItDESHandle.m]: Good catch! I fixed it in a commit... But now I have a bizarre issue where even on a default Ubuntu install, unar won't pass any password at all. Even a `printf("%d", length)` in this function yields 0. I could have sworn this was working a few weeks ago.

**Review by PaulTaykalo — APPROVED** [review]

**Comment by PaulTaykalo** [comment]
Thanks for the contribution!

**Pull request merged**